### ISSUE TITLE
Error while building Qt on linux system (red hat)

### ISSUE THREAD

**Issue by pramodk** [spack/spack#1517]
Hello All,

I see there are open issues/pull requests related to Qt package changes. Please close this ticket if this is already being addressed.

While building Qt on red hat 6.7 I see following errors:

```
spack install -v --keep-stage  qt@5.4.2 %gcc@4.9.3
```

Spack error:

```
/gpfs/bbp.cscs.ch/home/kumbhar/workarena/systems/lugviz/softwares/sources/spack/var/spack/repos/builtin/packages/qt/package.py:186, in install:
     184      def install(self, spec, prefix):
     185          self.configure()
  >> 186          make()
     187          make("install")
==> Error: Installation process had nonzero exit code : 256
```

Spack build log

```
                 from ../3rdparty/javascriptcore/JavaScriptCore/API/JSObjectRef.cpp:31:
../3rdparty/javascriptcore/JavaScriptCore/yarr/RegexJIT.h: In function ‘int QTJSC::Yarr::executeRegex(QTJSC::Yarr::RegexCodeBlock&, const UChar*, unsigned int, unsigned int, int*, int)’:
../3rdparty/javascriptcore/JavaScriptCore/yarr/RegexJIT.h:89:88: error: ‘jsRegExpExecute’ was not declared in this scope
         return (jsRegExpExecute(fallback, input, length, start, output, outputArraySize) < 0) ? -1 : output[0];
                                                                                        ^
make[3]: *** [obj/release/JSCallbackFunction.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[3]: *** [obj/release/JSObjectRef.o] Error 1
In file included from ../3rdparty/javascriptcore/JavaScriptCore/assembler/MacroAssembler.h:46:0,
                 from ../3rdparty/javascriptcore/JavaScriptCore/bytecode/Instruction.h:32,
                 from ../3rdparty/javascriptcore/JavaScriptCore/bytecode/CodeBlock.h:34,
                 from ../3rdparty/javascriptcore/JavaScriptCore/bytecode/CodeBlock.cpp:31:
../3rdparty/javascriptcore/JavaScriptCore/assembler/MacroAssemblerX86_64.h: In member function ‘QTJSC::AbstractMacroAssembler<QTJSC::X86Assembler>::Call QTJSC::MacroAssemblerX86_64::call()’:
../3rdparty/javascriptcore/JavaScriptCore/assembler/MacroAssemblerX86_64.h:113:22: warning: variable ‘label’ set but not used [-Wunused-but-set-variable]
         DataLabelPtr label = moveWithPatch(ImmPtr(0), scratchRegister);
                      ^
../3rdparty/javascriptcore/JavaScriptCore/assembler/MacroAssemblerX86_64.h: In member function ‘QTJSC::AbstractMacroAssembler<QTJSC::X86Assembler>::Call QTJSC::MacroAssemblerX86_64::tailRecursiveCall()’:
../3rdparty/javascriptcore/JavaScriptCore/assembler/MacroAssemblerX86_64.h:121:22: warning: variable ‘label’ set but not used [-Wunused-but-set-variable]
         DataLabelPtr label = moveWithPatch(ImmPtr(0), scratchRegister);
                      ^
../3rdparty/javascriptcore/JavaScriptCore/assembler/MacroAssemblerX86_64.h: In member function ‘QTJSC::AbstractMacroAssembler<QTJSC::X86Assembler>::Call QTJSC::MacroAssemblerX86_64::makeTailRecursiveCall(QTJSC::AbstractMacroAssembler<QTJSC::X86Assembler>::Jump)’:
../3rdparty/javascriptcore/JavaScriptCore/assembler/MacroAssemblerX86_64.h:130:22: warning: variable ‘label’ set but not used [-Wunused-but-set-variable]
         DataLabelPtr label = moveWithPatch(ImmPtr(0), scratchRegister);
                      ^
In file included from ../3rdparty/javascriptcore/JavaScriptCore/runtime/RegExp.h:30:0,
                 from ../3rdparty/javascriptcore/JavaScriptCore/bytecode/CodeBlock.h:40,
                 from ../3rdparty/javascriptcore/JavaScriptCore/bytecode/CodeBlock.cpp:31:
../3rdparty/javascriptcore/JavaScriptCore/yarr/RegexJIT.h: In destructor ‘QTJSC::Yarr::RegexCodeBlock::~RegexCodeBlock()’:
../3rdparty/javascriptcore/JavaScriptCore/yarr/RegexJIT.h:65:36: error: ‘jsRegExpFree’ was not declared in this scope
             jsRegExpFree(m_fallback);
                                    ^
In file included from ../3rdparty/javascriptcore/JavaScriptCore/runtime/RegExp.h:30:0,
                 from ../3rdparty/javascriptcore/JavaScriptCore/bytecode/CodeBlock.h:40,
                 from ../3rdparty/javascriptcore/JavaScriptCore/bytecode/CodeBlock.cpp:31:
../3rdparty/javascriptcore/JavaScriptCore/yarr/RegexJIT.h: In function ‘int QTJSC::Yarr::executeRegex(QTJSC::Yarr::RegexCodeBlock&, const UChar*, unsigned int, unsigned int, int*, int)’:
../3rdparty/javascriptcore/JavaScriptCore/yarr/RegexJIT.h:89:88: error: ‘jsRegExpExecute’ was not declared in this scope
         return (jsRegExpExecute(fallback, input, length, start, output, outputArraySize) < 0) ? -1 : output[0];
                                                                                        ^
make[3]: *** [obj/release/CodeBlock.o] Error 1
make[3]: Leaving directory `/tmp/kumbhar/spack-stage/spack-stage-T1p9sl/qt-everywhere-opensource-src-5.4.2/qtscript/src/script'
make[2]: *** [sub-script-make_first-ordered] Error 2
make[2]: Leaving directory `/tmp/kumbhar/spack-stage/spack-stage-T1p9sl/qt-everywhere-opensource-src-5.4.2/qtscript/src'
make[1]: *** [sub-src-make_first] Error 2
make[1]: Leaving directory `/tmp/kumbhar/spack-stage/spack-stage-T1p9sl/qt-everywhere-opensource-src-5.4.2/qtscript'
make: *** [module-qtscript-make_first] Error 2
```

Let me know if I can try something else to avoid the build failure.

**Comment by pramodk** [comment]
Note that disabling qtscript I can "avoid" this issue:

```
diff --git a/var/spack/repos/builtin/packages/qt/package.py b/var/spack/repos/builtin/packages/qt/package.py
index e496a3e..b717247 100644
--- a/var/spack/repos/builtin/packages/qt/package.py
+++ b/var/spack/repos/builtin/packages/qt/package.py
@@ -179,6 +179,7 @@ class Qt(Package):
                   # If someone wants to get a webkit build working, be my
                   # guest!
                   '-skip', 'qtwebkit',
+                  '-skip', 'qtscript',
```

**Comment by adamjstewart** [comment]
See https://github.com/LLNL/spack/pull/1041#issuecomment-238363551 and https://github.com/LLNL/spack/pull/1455#issuecomment-239507716. The problem is that qt depends on glib, and glib now depends on pcre. The pcre headers are messing with qt's headers. @xjrc and I are debating a solution for preventing qt from depending on pcre.

**Comment by pramodk** [comment]
thanks adam for info, it would be great to have qt build working. I will wait for the changes!

**Comment by svenevs** [comment]
@adamjstewart why can't we lock the versions when coming from qt as an intermediate solution? There are tons of other packages that do this, I think.

```py
# glib also has 2.42, why not grab that instead?
depends_on('pcre+utf', when='@2.48:')
```

Or alternatively, maybe making the system ones available through spack (some `~binutils`-esque shenanigans)?

**Comment by mjwoods** [comment]
@svenevs - I tried your suggestion of pinning `glib@2.42.1` to avoid the dependency on pcre. This worked for building qt. But it seems that various packages depend on pinned versions of glib - and unfortunately not the same version! Examples include gconf, gmime and gobject-introspection. Some of these packages are sometimes used together with qt, directly or indirectly - as in the case of metview, which caused spack to fail during the concretization phase.

**Comment by svenevs** [comment]
hey @mjwoods so the issue you describe is unavoidable.  as it turns out, in recent weeks i've taken a new tactic wrt to `qt` -- cheat around its dependencies.

### hacky solution

```py
# depends_on("glib", when='@4:')
# depends_on("gtkplus", when='+gtk')
```

Comment out the official `depends_on`, **keep the variants** (if you change the variants, the hash will change and you will get grief elsewhere).  This fools `spack` into thinking that it's installed `qt+gtkplus` (of course you had to explicitly ask `spack install +gtk`), but it never shows up in the dependencies.

**NOTE** that this is only valid in the `gtkplus` case because I have installed `gtk+-3.0` using my system package manager (`dnf`, `apt-get`, maybe `brew`?).

### example installation

For reference, I have `qt` installed as follows:

```console
$ spack find -ldfv qt
==> 1 installed packages.
-- linux-fedora23-x86_64 / gcc@5.3.1 ----------------------------
stgf7p6    qt@5.8.0%gcc~dbus~examples+gtk~krellpatch~mesa+phonon+webkit
uszwb52        ^fontconfig@2.12.1%gcc
i6knxgr            ^font-util@1.3.1%gcc
nhmdfwa            ^freetype@2.7.1%gcc
6mu4dee                ^bzip2@1.0.6%gcc+shared
yni2235                ^libpng@1.6.29%gcc
br6mxvn                    ^zlib@1.2.11%gcc+pic+shared
exxfg3q            ^libxml2@2.9.4%gcc~python
b5rfb4d                ^xz@5.2.3%gcc
55eakgr        ^gperf@3.0.4%gcc
xwursrz        ^icu4c@58.2%gcc
izuys7e        ^jpeg@9b%gcc
fvbdn4j        ^libmng@2.0.2%gcc
tg667bc            ^lcms@2.8%gcc
dtf6lto                ^libtiff@4.0.7%gcc
27yfhzi        ^libxcb@1.12%gcc
cgkm7vy            ^libpthread-stubs@0.4%gcc
zpzzo6z            ^libxau@1.0.8%gcc
wwc5mx3            ^libxdmcp@1.1.2%gcc
lyj3ary        ^openssl@1.0.2k%gcc
```

That is, there is no actual `glib` or `gtkplus` dep in my spec, making my `qt` able to solve your problem (yes...you're gonna have to recompile `qt` to do this...).

### why do this

In my experience, _even if you get `qt` to compile_, it just flat out won't work the way you want.  I fully understand that `spack` is intended for generally not me -- I am only administering my personal system, I don't have to worry about reproducible builds, lack of `root` capabilities to install dependencies, etc.  However, I've come to believe (with only experience, and no actual understanding), that installing _any_ version of `glibc`, and the "core" packages that relate to the operating system and its necessities, are unstable at best.  Basically, even when I got `qt` to compile, strange behavior and unexplained segfaults would ensue.  I believe this is because my desktop manager (`gnome`) and kernel were using one version `glibc`, and `qt` was using a different one.

#### option A

doesn't solve your problem of explicit `glibc` dependencies, but you can try and favor the correct version of `glibc` for your OS.

```console
$ /lib64/libc.so.6 
GNU C Library (GNU libc) stable release version 2.22, by Roland McGrath et al.
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 5.3.1 20160406 (Red Hat 5.3.1-6).
Available extensions:
	The C stubs add-on version 2.1.2.
	crypt add-on version 2.1 by Michael Glad and others
	GNU Libidn by Simon Josefsson
	Native POSIX Threads Library by Ulrich Drepper et al
	BIND-8.2.3-T5B
	RT using linux kernel aio
libc ABIs: UNIQUE IFUNC
For bug reporting instructions, please see:
<http://www.gnu.org/software/libc/bugs.html>.
```

so I would install `glibc@2.22` (you'd have to add your version btw), and using `packages.yaml` I would try and always force this version.  Of course, if a package asks for a specific version of `glib`, all bets are off.

#### option b

this is what I do now.  I meticulously run `spack spec -I` for any installation before running it.  If I see `glib` pop up I `spack edit` the package it comes from, and remove the dependency.  This works because ultimately my `/lib64` libraries will eventually get picked up, whether `spack` wants them to or not.

it's a dirty hack, but it's been successful for me :wink:

### summary

If you don't need to worry about deployment on multiple servers / only _actually_ need one version of `glib`, use the one that comes with your operating system.  The `glib` versions seem to sometimes correlate with other packages your system may have, but I know not enough to suggest any actual fixes...only hacks...

It's worth disclaiming that you are now putting yourself in a strange scenario where `spack` is concerned.  It's dependencies are now no longer self-contained.  YMMV but I hope it helps you get whatever you need built!

**Comment by adamjstewart** [comment]
> Comment out the official `depends_on`, **keep the variants** (if you change the variants, the hash will change and you will get grief elsewhere).

Actually, removing the dependencies will change the hash too. With the dependencies, the hash for me is `tqdqn3jmdzomno2bynjouj2jsgp6w6rs`. Without the dependencies, the hash is `mfn6ob6nk2qwhsyndbeidhlbuugoaaxn`. You can use `spack spec -L qt` to confirm this. Obviously your hashes will be different from mine since you're on a different OS.

In general, hacking packages is a bad idea. Assuming of course that you want to be able to update Spack without any woes. The non-hacky equivalent is telling Spack to use your system glib installation. Run `spack config edit packages` to open `~/.spack/packages.yaml` and add the following snippet:
```yaml
packages:
  glib:
    paths:
      glib@2.22: /
    buildable: False
  gtkplus:
    paths:
      gtkplus@2.24.25: /
    buildable: False
```
Replace the versions and directories with your system installed versions and directories. `buildable: False` tells Spack that it isn't allowed to install these packages even if a newer version is available. See the [External Packages](http://spack.readthedocs.io/en/latest/build_settings.html#external-packages) documentation for more information.

**Comment by mjwoods** [comment]
Hi @pramodk , I think your issue is resolved by #4270 , which has now been merged. If you update to the latest spack, it should be possible to install qt on your system.

**Comment by pramodk** [comment]
@mjwoods : if you have tested this we can close this issue. (for me it's failing due to other dependency errors, I need to figure out whats going on).

**Comment by alalazo** [comment]
@pramodk @pramodskumbhar Still interested in this? Is it solved or not?

**Comment by svenevs** [comment]
> Is it solved or not?

FWIW in recent attempts I've still had great pain with Qt.  I've resorted to always installing it with my system package manager and making it `buildable: false` in `packages.yaml`.  I probably got the variants right, but who knows.  The last time I tried I got different errors than I had seen in previous attempts, likely indicating a more intricate role of the compiler than is desireable (recent attempts were on an upgraded distribution with a new default compiler).  But I very much did not investigate...
